Implement flexible box layout for row container

stshine opened the pull request:
<!-- Please describe your changes on the following line: -->

This pull requests implements basic flexible box layout for row container.
It  has implemented most basic flexbox features, including grow, shrink, multi-line, *reverse properties, and alignment under `justify-content`, `align-items`, `align-self`, `align-content`.

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X]  There are tests for these changes

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

r? @pcwalton

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/12330)

<!-- Reviewable:end -->

stshine committed:
Extend fields of FlexItem and add methods

Extend fields of `FlexItem` struct with values that are necessary to
resolve flexible lengths, and the 'order' property. Add other methods
for size computing to make the code more modular.

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify layout code, but no tests are modified. Please consider adding a test!

stshine commented:
There a several concerns for this pull request.

One is that previously every method involves a vtable lookup, so I add a style field and clone the style of the flow to it, but I am not very sure what the price is.
Another one is that there are occurrences of 'box-sizing' property here and there to get the proper main size. I wonder if there is a way to reduce them.

stshine commented:
ping @pcwalton 
Mind share some general opinion about this pull request? There is a bunch of PRs pending on this.

shinglyu commented:
(tracking #12453)

shinglyu commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/layout/flex.rs, line 145 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12330#-KMwN6VAs86npoGM6RhZ:-KMwN6VAs86npoGM6Rh_:1563241048) ([raw file](https://github.com/servo/servo/blob/ee4755b9b9beb02744f657c2ea57a098329bbe29/components/layout/flex.rs#L145)):*
> ```Rust
> impl FlexItem {
>     pub fn new(flow: FlowRef) -> FlexItem {
>         let style = flow.as_block().fragment.style.clone();
> ```

Will the style ever change after we clone it (e.g. by JavaScript)? Do we need to sync it up with the style side?

---

*[components/layout/flex.rs, line 222 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12330#-KMwOZ5Jp9XAoGomfnEJ:-KMwOZ5Jp9XAoGomfnEK:1993642271) ([raw file](https://github.com/servo/servo/blob/ee4755b9b9beb02744f657c2ea57a098329bbe29/components/layout/flex.rs#L222)):*
> ```Rust
>         let ref fragment = self.flow.as_block().fragment;
>         let adjustment;
>         match mode {
> ```

We could just `let adjustment = match mode {...}`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12330)*
<!-- Sent from Reviewable.io -->

shinglyu commented:
Disclaimer: I don't have the right to r+ yet, just providing some feedback.

stshine commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[components/layout/flex.rs, line 145 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12330#-KMwN6VAs86npoGM6RhZ:-KMx3Q5LSHlaiyp-KMyX:-682205792) ([raw file](https://github.com/servo/servo/blob/ee4755b9b9beb02744f657c2ea57a098329bbe29/components/layout/flex.rs#L145)):*
<details><summary><i>Previously, shinglyu (Shing Lyu) wrote…</i></summary>
> Will the style ever change after we clone it (e.g. by JavaScript)? Do we need to sync it up with the style side?
</details>
What cloned is just a Arc pointer to the actual `ServoComputedValues`, so It should not be a problem.

---

*[components/layout/flex.rs, line 191 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12330#-KMxm5sNISQk1u8SsF9t:-KMxm5sO9s9ZF1V_sqNv:50195832) ([raw file](https://github.com/servo/servo/blob/ee4755b9b9beb02744f657c2ea57a098329bbe29/components/layout/flex.rs#L191)):*
> ```Rust
>                         (MaybeAuto::from_style(margin.inline_start, Au(0)).specified_or_zero() +
>                          MaybeAuto::from_style(margin.inline_end, Au(0)).specified_or_zero())
>                     }
> ```

Here when `box-sizing` is set to `border-box`, The border and padding length contained in main size is computed at intrinsic size calculating, which set percentage padding to zero. Maybe I should take percentage padding into account for inline main size.

---

*[components/layout/flex.rs, line 222 \[r1\]](https://reviewable.io:443/reviews/servo/servo/12330#-KMwOZ5Jp9XAoGomfnEJ:-KMx3ieTeOlxGQPXxyup:2091736419) ([raw file](https://github.com/servo/servo/blob/ee4755b9b9beb02744f657c2ea57a098329bbe29/components/layout/flex.rs#L222)):*
<details><summary><i>Previously, shinglyu (Shing Lyu) wrote…</i></summary>
> We could just `let adjustment = match mode {...}`
</details>
Previously I do the border and margin computing here. Will fix it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12330)*
<!-- Sent from Reviewable.io -->

stshine commented:
> Disclaimer: I don't have the right to r+ yet, just providing some feedback.
May I ask your general opinion about this pull request? If you think this pull request is acceptable, I will make other PRs that depend on this.

pcwalton commented on components/layout/flex.rs:
```diff
@@ -69,22 +73,195 @@ impl AxisSize {
+    /// The 'order' property of this item.
+    pub order: i32,
+    /// Whether the main size has met its constraint.
+    pub is_freezed: bool,
```
nit: should this be `is_frozen`?

pcwalton commented on components/layout/flex.rs:
```diff
@@ -69,22 +73,195 @@ impl AxisSize {
+    }
+
+    /// Return the outer main size of the item, including paddings and margins,
+    /// clamped by max and min size.
```
Maybe this should become a method on `Fragment`?

stshine replied:
All these methods exists as the result of refactoring the method of obtaining a line in flexbox. Both this and the adjustment issues remain here because I previously do margin and padding calculation in `outer_main_size()` . This point is quite valid, but 1. I am not sure if other components need this method, 2. This will make the style field unnecessary and make `auto_margin_num()` a function instead of a method. Maybe I should do this after more consideration,  together with renaming the `Mode` enum to `Direction` and moving it to `model.rs`.

pcwalton commented on components/layout/flex.rs:
```diff
@@ -69,22 +73,195 @@ impl AxisSize {
+    pub fn outer_main_size(&self, mode: Mode) -> Au {
+        let ref fragment = self.flow.as_block().fragment;
+        let adjustment;
+        match mode {
```
nit: you can shorten this to `let adjustment = match mode { ... }` and then remove the `adjustment =`s below and the semicolons

pcwalton commented on components/layout/flex.rs:
```diff
@@ -69,22 +73,195 @@ impl AxisSize {
+    }
+}
+
+/// Represent a child in a flex container. Most fields here are used in
```
nit: "Represent" -> "Represents"

pcwalton commented:
Looks good, just had a few nits. Thanks!

stshine commented on components/layout/flex.rs:
```diff
@@ -69,22 +73,195 @@ impl AxisSize {
+                        let margin = self.style.logical_margin();
+                        (MaybeAuto::from_style(margin.inline_start, Au(0)).specified_or_zero() +
+                         MaybeAuto::from_style(margin.inline_end, Au(0)).specified_or_zero())
+                    }
```
Here when the the `box-sizing` property is `border-box`, percentage padding included in main size is calculated to zero as a result of `bubble_inline_sizes`. Maybe I should recalculate them?

stshine commented:
Issues addressed. Thanks!

shinglyu commented:
Does any other code use these methods? If it's for future patches, maybe we should merge this right before the one that uses it.

stshine commented:
@shinglyu told me for a big feature I should make separate commits and submit them in on PR. If so I will close other PRs, rename this PR, and push consequent commits for flexbox here.

stshine committed:
Add flags 'is_wrappable' and 'cross_reverse'

These 'is_wrappable' flag can be used to determine if a flex flow can be
multi-line. The 'is_reverse' flag is replaced by 'main_reverse' and
'cross_reverse' since flex flow can go reverse individually in main and
cross direction.

stshine committed:
Add FlexLine and its method for flex resolving

Add the 'FlexLine' struct to represent a line in flex container, and the
'flex_resolve()' method for flex lengths resolving.

stshine committed:
Add 'get_flex_line()' method to flex flow

The method returns 'Option<FlexLine>', can be used to obtain a line in
flex container.

stshine committed:
Filter abs-pos children and sort by 'order' field

As the flexbox spec change in May 2016, the absolutely positioned
children painting order no longer follow the `order` property, thus we
can simply filter them out.

Also sort items by the order field of 'FlexItem', no longer do two
vtable lookups in each compare.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #12515) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
That sounds like a good plan.

stshine committed:
Make the width of block customizable by its parent

Add a 'IS_FLEX' flag to bitflags of block flow, and when it is set, the
block will use the inline size and margins set by its parent.

stshine committed:
Flex elements in a row flex container

Iterate over the items, create lines from them, resolve flex lengths
against them, set their inline size and margin lengths, and align them
according to 'justify-content', 'align-content' and 'align-self'.

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit a70d2bb with merge 6d1bacf...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2282)

stshine commented:
So this PR is ready. Through it has implement most basic features, There are still a lot of things missing . except the most import support for column container,  some missing features I gathered is:
- `bubble_inline_sizes()` is not implemented correctly;
- absolutely positioned children is not layouted yet;
- `visibility: hidden` is not implemented;
- `align-self: baseline` is not implemented;
- The interaction between block flow and flex flow needs huge enhancement.

However, IMO the core logic is stable enough for review. Since I am not  experienced, criticisms are very much appreciated!

pcwalton commented on components/layout/flex.rs:
```diff
@@ -69,22 +72,192 @@ impl AxisSize {
+                self.max_size = specified_or_none(self.style.max_inline_size(), containing_length)
+                    .unwrap_or(MAX_AU);
+                self.min_size = specified(self.style.min_inline_size(), containing_length);
+            },
```
uber-nit: no need for this `,`

pcwalton commented on components/layout/flex.rs:
```diff
@@ -69,22 +72,192 @@ impl AxisSize {
+                        block.fragment.border_padding.inline_start_end()
+                };
+                let content_size = block.base.intrinsic_inline_sizes.preferred_inline_size
+                    - block.fragment.surrounding_intrinsic_inline_size() + adjustment;
```
Can we factor this out into a method on block flows?

pcwalton commented on components/layout/flex.rs:
```diff
@@ -69,22 +72,192 @@ impl AxisSize {
+                    box_sizing::T::border_box =>
+                        fragment.margin.inline_start_end()
+                }
+            },
```
nit: no `,`

pcwalton commented on components/layout/flex.rs:
```diff
@@ -69,22 +72,192 @@ impl AxisSize {
+        max(self.min_size, min(self.base_size, self.max_size)) + adjustment
+    }
+
+    pub fn auto_margin_num(&self, mode: Mode) -> i32 {
```
nit: How about `auto_margin_count`? Could possibly add a comment saying that this counts the number of margins in the provided direction.

pcwalton commented on components/layout/flex.rs:
```diff
@@ -262,6 +263,98 @@ impl FlexItem {
+struct FlexLine {
+    /// Range of items belong to this line in 'self.items'.
+    pub range: Range<usize>,
+    /// Remainig free space of this line, items will grow or shrink based on it being positive or negative.
```
typo: "Remainig" -> "Remaining"

pcwalton commented on components/layout/flex.rs:
```diff
@@ -262,6 +263,98 @@ impl FlexItem {
+    pub range: Range<usize>,
+    /// Remainig free space of this line, items will grow or shrink based on it being positive or negative.
+    pub free_space: Au,
+    /// the number of auto margins of items.
```
uber-nit: capitalize "the"

pcwalton commented on components/layout/flex.rs:
```diff
@@ -421,6 +421,42 @@ impl FlexFlow {
+            0
+        } else {
+            self.lines[self.lines.len()-1].range.end
+        };
```
Could be just `let start = self.lines.last().map(|line| line.range.end).unwrap_or(0);`

pcwalton commented on components/layout/flex.rs:
```diff
@@ -421,6 +421,42 @@ impl FlexFlow {
+        for mut item in items.iter_mut() {
+            item.init_sizes(container_size, self.main_mode);
+            let outer_main_size = item.outer_main_size(self.main_mode);
+            if total_line_size + outer_main_size > container_size && end != start  && self.is_wrappable {
```
nit: just one space before the `&&`

pcwalton commented on components/layout/flex.rs:
```diff
@@ -421,6 +421,42 @@ impl FlexFlow {
+        let mut margin_count = 0;
+
+        let items = &mut self.items[start..];
+        for mut item in items.iter_mut() {
```
nit: could just be `for mut item in items`

pcwalton commented on components/layout/block.rs:
```diff
@@ -3057,6 +3077,30 @@ impl ISizeAndMarginsComputer for InlineBlockReplaced {
+                                                                container_block_size);
+    }
+
+    // Literally do nothing.
```
I would say something like "The inline size has already been set by our flex flow parent, so do nothing here."

pcwalton commented on components/layout/flex.rs:
```diff
@@ -581,30 +577,106 @@ impl FlexFlow {
+        let explicit_content_size = self
+                                    .block_flow
+                                    .explicit_block_size(parent_container_size)
+                                    .map(|x| if x < box_border { Au(0) } else { x - box_border });
```
Could just be `max(x - box_border, Au(0))`

pcwalton commented on components/layout/flex.rs:
```diff
@@ -581,30 +577,106 @@ impl FlexFlow {
+                    block.base.position.start.i =
+                        inline_start_content_edge *2 + content_inline_size
+                        - cur_i  - item_outer_size;
+                };
```
Strange formatting. Perhaps:

    block.base.position.start.i = if !self.main_reverse {
        cur_i
    } else {
        inline_start_content_edge * 2 + content_inline_size - cur_i - item_outer_size
    };

pcwalton commented on components/layout/flex.rs:
```diff
@@ -628,63 +700,137 @@ impl FlexFlow {
+        let explicit_content_size = self
+                                    .block_flow
+                                    .explicit_block_size(parent_container_size)
+                                    .map(|x| if x < box_border { Au(0) } else { x - box_border });
```
Could use `max()` here again

pcwalton commented on components/layout/flex.rs:
```diff
@@ -628,63 +700,137 @@ impl FlexFlow {
+                    if line_count == 1 {
+                        Au(0)
+                    } else {
+                        free_space /(line_count - 1)
```
uber-nit: space after `/`

pcwalton commented on components/layout/flex.rs:
```diff
@@ -628,63 +700,137 @@ impl FlexFlow {
+                    if !self.cross_reverse {
+                        cur_b
+                    } else {
+                        self.block_flow.fragment.border_padding.block_start *2
```
nit: space after `*`

stshine committed:
Rename enum 'Mode' to 'Direction'

Rename the 'Mode' enum to 'Direction' and move it to 'model.rs', thus it
can be used to indicate logical direction.

stshine committed:
Refator 'box-sizing' calculation into a method

The 'box_sizing_boundary()' method return the border_padding width in
certain direction if fragment has property 'box-sizing: border-box'. The
'border_padding' field should be initialized before.

stshine commented:
I refactored the 'box-sizing: border box' calculation into a method of `Fragment`. Other refactoring may be easier after make the `start_end*` methods of `LogicalMargin` accept a direction parameter.

Any suggestion on the naming?

stshine committed:
Add Minor fixes and tests

notriddle commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 117b9c3 with merge 166a2af...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2208)

highfive commented:
▶ FAIL [expected PASS] /css-flexbox-1_dev/html/flexbox_direction-row-reverse.htm
      └   → /css-flexbox-1_dev/html/flexbox_direction-row-reverse.htm feda5dd29dd30f708fa4ebf388bb3ff8c5a6259a
    /css-flexbox-1_dev/html/reference/flexbox_direction-row-reverse-ref.htm 03b5b004667cb9aaa6da36eec4c2547436b14890
    Testing feda5dd29dd30f708fa4ebf388bb3ff8c5a6259a == 03b5b004667cb9aaa6da36eec4c2547436b14890
    
      ▶ TIMEOUT [expected OK] /css-flexbox-1_dev/html/flexbox_computedstyle_min-width-auto.htm

pcwalton commented:
@stshine The naming you chose seems fine to me.

stshine commented:
Can't understand why the last letter 'L' moved a bit in `flexbox_direction-row-reverse'. Maybe some intermittent error in text layout?
http://i.imgur.com/UcQy10x.png

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit b2a8d74 with merge 49976e5...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2324)

shinglyu commented:
Sometimes the font stuff in the CI is different from running locally. Can that test pass when you run locally?

stshine commented:
@shinglyu I believe the last failure is an intermittent failure caused by #12552, and the flexbox tests should go correctly now, don't worry :laughing: 
Still not sure why that letter moves one pixel on mac specifically, maybe something related to subpixel.

shinglyu commented:
Yeah probably, sometimes different platform rounds appunit differently and will cause the text to shift.

shinglyu commented:
@stshine By the way, after you land this patch, would you mind opening some followup bugs for the missing features? If you can provide some detail about them, I believe some contributors will be able to jump in and help.

stshine commented:
@shinglyu  Since this is just a basic flexbox implementation,  I have been collecting issues for a while, among which the most significant one is the column container support. Implementing this and `align-self:stretch` get complicated because 1. how a block determine its widths/height need to be dramatically changed, 2. this contradicts heavily with the parallel layout mechanism of servo. 

  But I need to sort all info carefully in my head and dive more to the layout system more to state those issues properly and with more details. I will reply to the main issue soon after this get merged.

pcwalton commented:
This looks good to me. @stshine, let me know when you feel this is ready and I'll r=me. Sorry about taking so long here!

stshine commented:
@pcwalton Sure it is! No problem, I know you are quite busy recently :) Thank you very much for taking time to review!

pcwalton commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit b2a8d74 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton b2a8d74237483abed9d72c493a4905fd1c098825 -->

KiChjang commented:
@bors-servo try- clean retry r=pcwalton

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.

bors-servo commented:
:pushpin: Commit b2a8d74 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton b2a8d74237483abed9d72c493a4905fd1c098825 -->

bors-servo commented:
:hourglass: Testing commit b2a8d74 with merge fe47bdf...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2334)

highfive commented:
▶ FAIL [expected PASS] /css-flexbox-1_dev/html/flexbox-flex-direction-default.htm
      └   → /css-flexbox-1_dev/html/flexbox-flex-direction-default.htm a505f9f57c0b83f64a1ebf31bb8fad4649d94d07
    /css-flexbox-1_dev/html/reference/flexbox-flex-direction-ref.htm 54fcc475c3f793d40ae0710e0c82021a8495aa35
    Testing a505f9f57c0b83f64a1ebf31bb8fad4649d94d07 == 54fcc475c3f793d40ae0710e0c82021a8495aa35
    
      ▶ FAIL [expected PASS] /css-flexbox-1_dev/html/flexbox-flex-direction-row-reverse.htm
      └   → /css-flexbox-1_dev/html/flexbox-flex-direction-row-reverse.htm a505f9f57c0b83f64a1ebf31bb8fad4649d94d07
    /css-flexbox-1_dev/html/reference/flexbox-flex-direction-ref.htm 54fcc475c3f793d40ae0710e0c82021a8495aa35
    Testing a505f9f57c0b83f64a1ebf31bb8fad4649d94d07 == 54fcc475c3f793d40ae0710e0c82021a8495aa35
    
      ▶ FAIL [expected PASS] /css-flexbox-1_dev/html/flexbox-flex-direction-row.htm
      └   → /css-flexbox-1_dev/html/flexbox-flex-direction-row.htm a505f9f57c0b83f64a1ebf31bb8fad4649d94d07
    /css-flexbox-1_dev/html/reference/flexbox-flex-direction-ref.htm 54fcc475c3f793d40ae0710e0c82021a8495aa35
    Testing a505f9f57c0b83f64a1ebf31bb8fad4649d94d07 == 54fcc475c3f793d40ae0710e0c82021a8495aa35
    
      ▶ FAIL [expected PASS] /css-flexbox-1_dev/html/flexbox-flex-wrap-wrap.htm
      └   → /css-flexbox-1_dev/html/flexbox-flex-wrap-wrap.htm a505f9f57c0b83f64a1ebf31bb8fad4649d94d07
    /css-flexbox-1_dev/html/reference/flexbox-flex-direction-ref.htm 54fcc475c3f793d40ae0710e0c82021a8495aa35
    Testing a505f9f57c0b83f64a1ebf31bb8fad4649d94d07 == 54fcc475c3f793d40ae0710e0c82021a8495aa35
    
      ▶ FAIL [expected PASS] /css-flexbox-1_dev/html/flexbox-flex-wrap-wrap-reverse.htm
      └   → /css-flexbox-1_dev/html/flexbox-flex-wrap-wrap-reverse.htm a505f9f57c0b83f64a1ebf31bb8fad4649d94d07
    /css-flexbox-1_dev/html/reference/flexbox-flex-direction-ref.htm 54fcc475c3f793d40ae0710e0c82021a8495aa35
    Testing a505f9f57c0b83f64a1ebf31bb8fad4649d94d07 == 54fcc475c3f793d40ae0710e0c82021a8495aa35
    
      ▶ FAIL [expected PASS] /css-flexbox-1_dev/html/flexbox_box-clear.htm
      └   → /css-flexbox-1_dev/html/flexbox_box-clear.htm c032f9ccd110b4e972d0da4d7a9f2d0f6398a5f4
    /css-flexbox-1_dev/html/reference/flexbox_box-clear-ref.htm eed758e56c9cd89490d0fdfda6531af771783e28
    Testing c032f9ccd110b4e972d0da4d7a9f2d0f6398a5f4 == eed758e56c9cd89490d0fdfda6531af771783e28

stshine commented:
Some upstream changes break the margin collapsing of references, which is sad. (Maybe also #12696 and #12676)

Wafflespeanut commented:
@bors-servo r=pcwalton

bors-servo commented:
:pushpin: Commit 60e2f44 has been approved by `pcwalton`

<!-- @bors-servo r=pcwalton 60e2f44a1950559398fa25aa9aa36cd2a7fa39ad -->

bors-servo commented:
:hourglass: Testing commit 60e2f44 with merge 15947f8...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/2272), [arm64](http://build.servo.org/builders/arm64/builds/2278), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2325), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2452), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2357), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2339), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2414), [windows-dev](http://build.servo.org/builders/windows-dev/builds/223)

pcwalton commented:
🎉 Thanks for pushing this through, @stshine!

stshine commented:
I feel satisfied and honored to be able to implement this for servo. Thank you @pcwalton again and thanks to everyone who helped!
